File: BinaryHeap/BinaryHeap.go
package main
import "fmt"

type BinaryHeap struct {
	list []int

}

func (heap *BinaryHeap) buildHeap() {
	for i := (len(heap.list) - 1) / 2; i > 0; i-- {
		heap.percolateDown(i)
	}
}

func (heap *BinaryHeap) percolateUp(hole int) {
	temp := heap.list[hole]

	for ; hole > 1 && temp < heap.list[hole / 2] ; hole /= 2 {
		heap.list[hole] = heap.list[hole / 2]
	}
	heap.list[hole] = temp
}

func (heap *BinaryHeap) percolateDown(hole int) {
	var child int
	temp := heap.list[hole]

	for ; hole * 2 <= len(heap.list) - 1; hole = child {
		child = hole * 2
		if (child != len(heap.list) - 1 && heap.list[child + 1] < heap.list[child]) {
			child++
		}

		if (heap.list[child] < temp) {
			heap.list[hole] = heap.list[child]
		} else {
			break
		}
	}
	heap.list[hole] = temp
}

func (heap *BinaryHeap) initializeEmptyHeap() {
	heap.list = make([]int, 1)
}

func (heap *BinaryHeap) initializeHeap(arr []int) {
	heap.list = make([]int, len(arr) + 1)
	for i := 0; i < len(arr); i++ {
		heap.list[i + 1] = arr[i]
	}
	heap.buildHeap()
}

func (heap *BinaryHeap) insert(value int) {
	if (cap(heap.list) > len(heap.list)) {
		heap.list = heap.list[0: len(heap.list) + 1]
		heap.list[len(heap.list) - 1] = value;
	} else {
		heap.list = append(heap.list, value)
	}
	heap.percolateUp(len(heap.list) - 1)
}

func (heap *BinaryHeap) removeMin() {
	if (len(heap.list) == 1) {
		fmt.Println("Sorry, but the heap is empty")

	} else if (len(heap.list) == 2) {
		heap.list = make([]int, 1)
	} else {
		heap.list[1] = heap.list[len(heap.list) - 1]
		heap.list = heap.list[0:len(heap.list) - 1]
		heap.percolateDown(1)
	}
}

func (heap *BinaryHeap) getMin() int {
	return heap.list[1]
}

func (heap *BinaryHeap) print() {
	for i := 1; i < len(heap.list); i++ {
		fmt.Print(heap.list[i], " ")
	}
	fmt.Println()
}

func main() {
	myHeap := BinaryHeap{}
	myHeap.initializeHeap([]int{7, 6, 5, 4})
}

/*
The capacity of a slice is the number of elements in 
the underlying array, counting from the first element in the slice.
*/

File: BinaryHeap/BinaryHeap.c
#include <stdio.h>
#include <stdlib.h>

struct BinaryHeap {
	int* array;
	int currentSize;
	int capacity;
};

void percolateDown(struct BinaryHeap* heap, int hole) {
	int child;
	int temp = heap->array[hole];

	for (; hole * 2 <= heap->currentSize; hole = child) {
		child = hole * 2;
		if (child != heap->currentSize && heap->array[child + 1] < heap->array[child]) {
			++child;
		}
		if (heap->array[child] < temp) {
			heap->array[hole] = heap->array[child];
		}
		else {
			break;
		}
	}
	heap->array[hole] = temp;
}

void percolateUp(struct BinaryHeap* heap, int hole) {
	int temp = heap->array[hole];

	for (; hole > 1 && temp < heap->array[hole / 2]; hole /= 2) {
		heap->array[hole] = heap->array[hole / 2];
	}
	heap->array[hole] = temp;
}

void buildHeap(struct BinaryHeap* heap) {
	for (int i = heap->currentSize / 2; i > 0; i--) {
		percolateDown(heap, i);
	}
}

void initializeHeap(struct BinaryHeap* heap, int* arr, int size) {
	heap->array = (int*)malloc(size * 2 * sizeof(int));
	for (int i = 0; i < size; i++) {
		heap->array[i + 1] = arr[i];
	}
	heap->currentSize = size;
	heap->capacity = size * 2;
	buildHeap(heap);
}

void initializeEmptyHeap(struct BinaryHeap* heap) {
	heap->array = malloc(10 * sizeof(int));
	heap->currentSize = 0;
	heap->capacity = 10;
}

void insert(struct BinaryHeap* heap, int value) {
	if (heap->currentSize + 1 >= heap->capacity) {
		heap->array = (int*)realloc(heap->array, heap->capacity * 2);
		heap->capacity *= 2;
	}
	heap->array[heap->currentSize + 1] = value;
	heap->currentSize++;
	percolateUp(heap, heap->currentSize);
}

void removeMin(struct BinaryHeap* heap) {
	if (heap->currentSize == 0) {
		printf("Sorry, but the heap is empty");
		return;
	}
	else if (heap->currentSize == 1) {
		heap->currentSize--;
	}
	else {
		heap->array[1] = heap->array[heap->currentSize];
		heap->currentSize--;
		percolateDown(heap, 1);
	}
}

int getMin(struct BinaryHeap* heap) {
	return heap->array[1];
}

void print(struct BinaryHeap* heap) {
	for (int i = 1; i <= heap->currentSize; i++) {
		printf("%d ", heap->array[i]);
	}
	printf("\n");
}

int main() {
	int initials[13] = {4, 3, 2, 3, 7, 89, 23, 14, 56, 23, 2, 1, 1};
	struct BinaryHeap* heap = malloc(sizeof(struct BinaryHeap));
	initializeHeap(heap, initials, 13);
}



/*
you need to store your struct on the heap in order
to malloc its array fields.
*/